1) Import Libraries
2) Import Data
3) Exercise
    - Mixed type Data
    - Missing values
        - Created new dataframe without missing values
    - Duplicates
        - Create new dataframe with no duplicates
    - Exported new clean dataframe
4) Task
    - Check all numeric columns check out with df.describe()
    - Check for mixed type data
    - Check for null values
    - Check for duplicate values
    - Export clean new dataframe

# 01 Import Libraries

In [2]:
#Import libraries
import pandas as pd
import numpy as np
import os

# 02 Importing Data

In [3]:
path = r'C:\Users\benar\CareerFoundry\09-2023 Instacart Basket Analysis'

In [4]:
df_prods = pd.read_csv(os.path.join(path, 'Data', 'Original Data', 'products.csv'))

In [4]:
df_ords = pd.read_csv(os.path.join(path, 'Data', 'Prepared Data', 'orders_wrangled.csv'))

# 03 Exercise

## Mixed Type Data

In [5]:
# Create dataframe
df_test = pd.DataFrame()

In [6]:
# Create mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [7]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [8]:
# Checking for mixed type columns (for loop)
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [9]:
# Change data types
df_test['mix'] = df_test['mix'].astype('str')

## Missing Values

In [10]:
# Search for missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [11]:
#Create dataframe with only missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [12]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

Replace missing values with mean
df['column with missings'].fillna(mean value, inplace=True)

Replace missing values with median
df['column with missings'].fillna(median value, inplace=True)

In [13]:
df_prods.shape

(49693, 5)

In [22]:
#Looking for product_id in prods dataframe to replace missing values
print(df_prods.loc[df_prods['product_id'].isin([34, 69, 116, 262, 525, 1511, 1780, 2240, 2586, 3159, 3230, 3736, 4283, 4790, 38183, 40440])])

       product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39

In [14]:
# Creating new dataframe without missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [15]:
df_prods_clean.shape

(49677, 5)

## Duplicates

In [16]:
# Creating dataframe with duplicate rows
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [17]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [18]:
df_prods_clean.shape

(49677, 5)

In [19]:
# Creating dataframe with no duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [20]:
df_prods_clean_no_dups.shape

(49672, 5)

In [27]:
# Exporting completed file
df_prods_clean_no_dups.to_csv(os.path.join(path, 'Data','Prepared Data', 'products_checked.csv'))

# 04 Task

In [22]:
#2 Run the df.describe() function on your df_ords dataframe
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

All looks good with min and max values. Nothing looks off, and means relatively check out for orders_day_of_week and orde_hour_of_day.

In [23]:
#3 Check for mixed-type data in your df_ords dataframe
for col2 in df_ords.columns.tolist():
  weird2 = (df_ords[[col2]].applymap(type) != df_ords[[col2]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird2]) > 0:
    print (col2)

#4 No column was printed, so there is no mixed type data.

In [24]:
#5 Run a check for missing values in your df_ords dataframe
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

days_since_prior_order makes sense to have missing values because any customer at one point is a new customer and so it's been 0 days since their prior order. It matches also that 206209 is the number of missing values and also is the max number of user_id results, so there isn't an inconistency there either.

#6 No missing values need addressing because these "missing" results make sense and shouldn't be changed, as explained above

In [25]:
#7 Run a check for duplicate values in your df_ords data
df_ords_dups = df_ords[df_ords.duplicated()]

In [26]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

No results in the new dataframe suggests no duplicates across rows

#8 No duplicates found

In [28]:
#9 Export clean files
df_ords.to_csv(os.path.join(path, 'Data','Prepared Data', 'orders_checked.csv'))

Products file exported above before Task